rust/src/ike/ike.rs

```diff
@@ -0,0 +1,628 @@
+            5 => Some(IkeEvent::WeakCryptoAuth),
+            6 => Some(IkeEvent::WeakCryptoNoDH),
+            7 => Some(IkeEvent::WeakCryptoNoAuth),
+            8 => Some(IkeEvent::InvalidProposal),
+            9 => Some(IkeEvent::UnknownProposal),
+            10 => Some(IkeEvent::PayloadExtraData),
+            _ => None,
+        }
+    }
+}
+
+pub struct IkeHeaderWrapper {
+    pub spi_initiator: String,
+    pub spi_responder: String,
+    pub ikev1_transforms: Vec<Vec<SaAttribute>>,
+    pub ikev2_transforms: Vec<Vec<IkeV2Transform>>,
+    pub ikev1_header: IkeV1Header,
+    pub ikev2_header: IkeV2Header,
+}
+
+impl IkeHeaderWrapper {
+    pub fn new() -> IkeHeaderWrapper {
+        IkeHeaderWrapper {
+            spi_initiator: String::new(),
+            spi_responder: String::new(),
+            ikev1_transforms: Vec::new(),
+            ikev2_transforms: Vec::new(),
+            ikev1_header: IkeV1Header::default(),
+            ikev2_header: IkeV2Header {
+                init_spi: 0,
+                resp_spi: 0,
+                next_payload: IkePayloadType::NoNextPayload,
+                maj_ver: 0,
+                min_ver: 0,
+                exch_type: IkeExchangeType(0),
+                flags: 0,
+                msg_id: 0,
+                length: 0
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct IkePayloadWrapper {
+    pub ikev1_payload_types: Option<HashSet<u8>>,
+    pub ikev2_payload_types: Vec<IkePayloadType>,
+}
+
+pub struct IKETransaction {
+    tx_id: u64,
+    progress: i32,
+
+    pub ike_version: u8,
+    pub hdr: IkeHeaderWrapper,
+    pub payload_types: IkePayloadWrapper,
+    pub notify_types: Vec<NotifyType>,
+
+    /// errors seen during exchange
+    pub errors: u32,
+
+    logged: LoggerFlags,
+    de_state: Option<*mut core::DetectEngineState>,
+    events: *mut core::AppLayerDecoderEvents,
+    tx_data: applayer::AppLayerTxData,
+}
+
+impl IKETransaction {
+    pub fn new() -> IKETransaction {
+        IKETransaction {
+            tx_id: 0,
+            progress: 0,
+            ike_version: 0,
+            hdr: IkeHeaderWrapper::new(),
+            payload_types: Default::default(),
+            notify_types: vec![],
+            logged: LoggerFlags::new(),
+            de_state: None,
+            events: std::ptr::null_mut(),
+            tx_data: applayer::AppLayerTxData::new(),
+            errors: 0
+        }
+    }
+
+    pub fn free(&mut self) {
+        if self.events != std::ptr::null_mut() {
+            core::sc_app_layer_decoder_events_free_events(&mut self.events);
+        }
+        if let Some(state) = self.de_state {
+            core::sc_detect_engine_state_free(state);
+        }
+    }
+
+    pub fn set_progress(&mut self, direction: u8) {
+        let mut progress = 1; // set to 1 to log and initiator_spi alert
+        if self.hdr.ikev1_header.key_exchange.len() > 0 && direction == STREAM_TOSERVER {
+            progress = 3;
```

Comment:
Could we have a name for this magic 3 constant ?

Reply:
And reuse this name in the C code for logging cf `SshStateBannerDone`

Reply:
checked on the whole progress feature and will drop it completely (always keep 1), the change a few PR's back of pushing alert-values into transactions and not alerting from state-values solved the problems i had...